File: 0572-subtree-of-another-tree/0572-subtree-of-another-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isSubtree(self, root: Optional[TreeNode], subRoot: Optional[TreeNode]) -> bool:
        stack = [root]
        
        def isSame(n1, n2) -> bool:
            if not n1 and not n2:
                return True
            elif not n1 or not n2:
                return False
            elif n1.val != n2.val:
                return False
            
            return isSame(n1.left, n2.left) and isSame(n1.right, n2.right)

        while stack:
            node = stack.pop()
            
            if isSame(subRoot, node):
                return True
            
            if node.left:
                stack.append(node.left)
            if node.right:
                stack.append(node.right)
            
        return False

File: 0572-subtree-of-another-tree/0572-subtree-of-another-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    /*boolean check(TreeNode a, TreeNode b){
        if(a == null && b == null) return true;
        else if(a == null || b == null || a.val != b.val) return false;
        
        return check(a.left, b.left) && check(a.right, b.right);
    }
    
    public boolean isSubtree(TreeNode root, TreeNode subRoot) {
        if(root == null) return false;
        if(root.val == subRoot.val && check(root, subRoot))
            return true;
        
        return isSubtree(root.left, subRoot) || isSubtree(root.right, subRoot);
    }*/
    public boolean isSubtree(TreeNode s, TreeNode t) {
        if (s == null) return false;
        if (isSame(s, t)) return true;
        return isSubtree(s.left, t) || isSubtree(s.right, t);
    }
    
    private boolean isSame(TreeNode s, TreeNode t) {
        if (s == null && t == null) return true;
        if (s == null || t == null) return false;
        
        if (s.val != t.val) return false;
        
        return isSame(s.left, t.left) && isSame(s.right, t.right);
    }
}